oxidation/libparsec/crates/core/src/invite/claimer.rs

```diff
@@ -0,0 +1,538 @@
+
+impl UserClaimInProgress2Ctx {
+    pub fn claimer_sas(&self) -> &SASCode {
+        &self.0.claimer_sas
+    }
+
+    pub async fn do_wait_peer_trust(self) -> InviteResult<UserClaimInProgress3Ctx> {
+        self.0
+            .do_wait_peer_trust()
+            .await
+            .map(UserClaimInProgress3Ctx)
+    }
+}
+
+#[derive(Debug)]
+pub struct DeviceClaimInProgress2Ctx(BaseClaimInProgress2Ctx);
+
+impl DeviceClaimInProgress2Ctx {
+    pub fn claimer_sas(&self) -> &SASCode {
+        &self.0.claimer_sas
+    }
+
+    pub async fn do_wait_peer_trust(self) -> InviteResult<DeviceClaimInProgress3Ctx> {
+        self.0
+            .do_wait_peer_trust()
+            .await
+            .map(DeviceClaimInProgress3Ctx)
+    }
+}
+
+#[derive(Debug)]
+pub struct UserClaimInProgress3Ctx(BaseClaimInProgress3Ctx);
+
+impl UserClaimInProgress3Ctx {
+    pub async fn do_claim_user(
+        self,
+        requested_device_label: Option<DeviceLabel>,
+        requested_human_handle: Option<HumanHandle>,
+    ) -> InviteResult<LocalDevice> {
+        // User&device keys are generated here and kept in memory until the end of
+        // the enrollment process. This mean we can lost it if something goes wrong.
+        // This has no impact until step 4 (somewhere between data exchange and
+        // confirmation exchange steps) where greeter upload our certificates in
+        // the server.
+        // This is considered acceptable given 1) the error window is small and
+        // 2) if this occurs the inviter can revoke the user and retry the
+        // enrollment process to fix this
+        let private_key = PrivateKey::generate();
+        let signing_key = SigningKey::generate();
+
+        let payload = InviteUserData {
+            requested_device_label,
+            requested_human_handle,
+            public_key: private_key.public_key(),
+            verify_key: signing_key.verify_key(),
+        }
+        .dump_and_encrypt(&self.0.shared_secret_key);
+
+        let rep = self
+            .0
+            .cmds
+            .send(invite_4_claimer_communicate::Req { payload })
+            .await?;
+
+        match rep {
+            invite_4_claimer_communicate::Rep::AlreadyDeleted => {
+                return Err(InviteError::AlreadyUsed)
+            }
+            invite_4_claimer_communicate::Rep::InvalidState => return Err(InviteError::PeerReset),
+            invite_4_claimer_communicate::Rep::NotFound => return Err(InviteError::NotFound),
+            invite_4_claimer_communicate::Rep::Ok { .. } => (),
+            invite_4_claimer_communicate::Rep::UnknownStatus { unknown_status, .. } => {
+                return Err(InviteError::Custom(format!(
+                    "Backend error during step 4 (data exchange): {unknown_status}"
+                )))
+            }
+        }
+
+        let rep = self
+            .0
+            .cmds
+            .send(invite_4_claimer_communicate::Req { payload: vec![] })
```

Comment:
Why do we send an empty `payload` ?

Reply:
This is something specific of the step 4:
This step is called communicate because it allow claimer and greeter to exchange messages (by opposition of the previous steps that were aimed at building the communication channel)

So the communicate works by having each peer sending a payload and receiving the other peer's payload. Here we have nothing to send, but want to receive the payload of the other peer. Hence sending an empty payload.